clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1407,16 +1401,16 @@ Sema::DeviceDiagBuilder Sema::SYCLDiagIfDeviceCode(SourceLocation Loc,
          "Should only be called during SYCL compilation");
   FunctionDecl *FD = dyn_cast<FunctionDecl>(getCurLexicalContext());
   DeviceDiagBuilder::Kind DiagKind = [this, FD] {
-    if (ConstructingOpenCLKernel)
+    if (ConstructingOpenCLKernel || !FD)
```

Comment:
FD can be NULL in the cases where getCurLexicalContext is TU/namespace/record/etc.  Don't we want to diagnose always in these cases, since they are likely static initializers?  Something like:

static auto Global = throw 3;

Reply:
I think we don't want to diagnose things like static initializers always (if they are not completely invalid for both host and device, but this case is diagnosed out of the box, I suppose).
Yes, I understand that static initializes run before main in regular C++, but there is no main in SYCL device code, only kernels and everything that used by kernels (functions/variables).
We want to diagnose only things that can get into device code. Things like static initializers shouldn't get into device code at all due to SYCL specific. SYCL spec even says: every variable with static storage duration that are odr-used inside a kernel must be const or constexpr and zero-initialized or constant-initialized.
I think all these rules tell that static initializers aren't executed on device and variables initialized in such way cannot be used inside the device code.

In addition, if we will diagnose such cases always, it will prevent from writing regular c++ code for host...

I think we want to diagnose case with in-class initializers, and in this case getCurLexicalContext is record, but we also want to diagnose it with deferred diagnostic because diagnostic is only needed if corresponding type is used in device code, but it requires changes in common parts of deferred diagnostics system, I'm looking into it https://github.com/intel/llvm/issues/1032 ...